Tools/build/bundle_ensurepip_wheels.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
+
+"""
+Download wheels for 'ensurepip' packages from PyPI.
+
+When GitHub Actions executes the script, output is formatted accordingly.
+https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-a-notice-message
+"""
+
+import importlib.util
+import os
+import sys
+from hashlib import sha256
+from pathlib import Path
+from urllib.error import URLError
+from urllib.request import urlopen
+
+HOST = 'https://files.pythonhosted.org'
+ENSURE_PIP_ROOT = Path(__file__, "..", "..", "..", "Lib", "ensurepip").resolve()
+WHEEL_DIR = ENSURE_PIP_ROOT / "_bundled"
+ENSURE_PIP_INIT = ENSURE_PIP_ROOT / "__init__.py"
+GITHUB_ACTIONS = "GITHUB_ACTIONS" in os.environ
+
+
+def print_notice(message: str) -> None:
+    if GITHUB_ACTIONS:
+        print(f"::notice::{message}", end="\n\n")
+    else:
+        print(message, file=sys.stderr)
+
+
+def print_error(message: str) -> None:
+    if GITHUB_ACTIONS:
+        print(f"::error::{message}", end="\n\n")
+    else:
+        print(message, file=sys.stderr)
+
+
+def download_wheels() -> int:
+    """Download wheels into bundle if they are not there yet."""
+
+    try:
+        projects = _get_projects()
+    except (AttributeError, TypeError):
+        print_error(f"Could not find '_PROJECTS' in {ENSURE_PIP_INIT}.")
+        return 1
+
+    errors = 0
+    for name, version, checksum in projects:
```

Comment:
FWIW with https://github.com/python/cpython/pull/109245, looping will stop making sense here. So maybe you could start working with just `projects[0]` already?

Reply:
No, let's keep them separate.

Reply:
I was kinda assuming that the simplification PR would get merged first, in which case, this one would have to adapt..